derp/derp_test.go

```diff
@@ -851,125 +851,189 @@ func TestClientSendPong(t *testing.T) {
+			t.Errorf("client %q isDup = %v; want %v", clientName[c], got, want)
 		}
-		return c
 	}
-
-	wantVar := func(v *expvar.Int, want int64) {
+	checkDisabled := func(c *sclient, want bool) {
 		t.Helper()
-		if got := v.Value(); got != want {
-			t.Errorf("got %d; want %d", got, want)
+		if got := c.isDisabled.Get(); got != want {
+			t.Errorf("client %q isDisabled = %v; want %v", clientName[c], got, want)
 		}
 	}
-
-	wantClosed := func(c *Client) {
+	wantDupConns := func(want int) {
 		t.Helper()
-		for {
-			m, err := c.Recv()
-			if err != nil {
-				t.Logf("got expected error: %v", err)
-				return
-			}
-			switch m.(type) {
-			case ServerInfoMessage:
-				continue
-			default:
-				t.Fatalf("client got %T; wanted an error", m)
-			}
+		if got := s.dupClientConns.Value(); got != int64(want) {
+			t.Errorf("dupClientConns = %v; want %v", got, want)
 		}
 	}
-
-	c1 := connect()
-	waitConnect(t, c1)
-	c2 := connect()
-	waitConnect(t, c2)
-	wantVar(&s.clientsReplaced, 1)
-	wantClosed(c1)
-
-	for i := 0; i < 100+5; i++ {
-		c := connect()
-		defer c.nc.Close()
-		if s.clientsReplaceLimited.Value() == 0 && i < 90 {
-			continue
+	wantDupKeys := func(want int) {
+		t.Helper()
+		if got := s.dupClientKeys.Value(); got != int64(want) {
+			t.Errorf("dupClientKeys = %v; want %v", got, want)
 		}
-		t.Logf("for %d: replaced=%d, limited=%d, sleeping=%d", i,
-			s.clientsReplaced.Value(),
-			s.clientsReplaceLimited.Value(),
-			s.clientsReplaceSleeping.Value(),
-		)
 	}
 
-	mu.Lock()
-	defer mu.Unlock()
-	if sleeps == 0 {
-		t.Errorf("no sleeps")
-	}
-	if slept == 0 {
-		t.Errorf("total sleep duration was 0")
-	}
+	// Common case: a single client comes and goes, with no dups.
+	run("one_comes_and_goes", func(t *testing.T) {
+		wantNoClient()
+		s.registerClient(c1)
+		wantSingleClient(c1)
+		s.unregisterClient(c1)
+		wantNoClient()
+	})
+
+	// A still somewhat common case: a single client was
+	// connected and then their wifi dies or laptop closes
+	// or they switch networks and connect from a
+	// different network. They have two connections but
+	// it's not very bad. Only their new one is
+	// active. The last one, being dead, doesn't send and
+	// thus the new one doesn't get disabled.
+	run("small_overlap_replacement", func(t *testing.T) {
+		wantNoClient()
+		s.registerClient(c1)
+		wantSingleClient(c1)
+		wantActive(c1)
+		wantDupKeys(0)
+		wantDupKeys(0)
+
+		s.registerClient(c2) // wifi dies; c2 replacement connects
+		wantDupSet()
+		wantDupConns(2)
+		wantDupKeys(1)
+		checkDup(c1, true)
+		checkDup(c2, true)
+		checkDisabled(c1, false)
+		checkDisabled(c2, false)
+		wantActive(c2) // sends go to the replacement
+
+		s.unregisterClient(c1) // c1 finally times out
+		wantSingleClient(c2)
+		checkDup(c2, false) // c2 is longer a dup
+		wantActive(c2)
+		wantDupConns(0)
+		wantDupKeys(0)
+	})
+
+	// Key cloning situation with concurrent clients, both trying
+	// to write.
+	run("concurrent_dups_get_disabled", func(t *testing.T) {
```

Comment:
Add a case with 3 clients that communicate ABCA, to verify we detect that too?

Reply:
Done.